flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/rank/UpdateRankFunction.java

```diff
@@ -0,0 +1,498 @@
+								// match to the previous position
+								return;
+							}
+
+							BaseRow rowKey = rowKeyIter.next();
+							RankRow prevRow = rowKeyMap.get(rowKey);
+							retract(out, prevRow.row, curRank - 1);
+							collect(out, prevRow.row, curRank);
+						}
+					} else {
+						// current sort key is smaller than old sort key, the rank is not changed, so skip
+						return;
+					}
+				}
+			} else {
+				curRank += rowKeys.size();
+			}
+		}
+
+		// remove the records associated to the sort key which is out of topN
+		List<BaseRow> toDeleteSortKeys = new ArrayList<>();
+		while (iterator.hasNext()) {
+			Map.Entry<BaseRow, Collection<BaseRow>> entry = iterator.next();
+			Collection<BaseRow> rowKeys = entry.getValue();
+			Iterator<BaseRow> rowKeyIter = rowKeys.iterator();
+			while (rowKeyIter.hasNext()) {
+				BaseRow rowKey = rowKeyIter.next();
+				rowKeyMap.remove(rowKey);
+				dataState.remove(rowKey);
+			}
+			toDeleteSortKeys.add(entry.getKey());
+		}
+		for (BaseRow toDeleteKey : toDeleteSortKeys) {
+			buffer.removeAll(toDeleteKey);
+		}
+	}
+
+	private void processElementWithoutRowNumber(BaseRow inputRow, Collector<BaseRow> out) throws Exception {
+		BaseRow sortKey = sortKeySelector.getKey(inputRow);
+		BaseRow rowKey = rowKeySelector.getKey(inputRow);
+		if (rowKeyMap.containsKey(rowKey)) {
+			// it is an updated record which is in the topN, in this scenario,
+			// the new sort key must be higher than old sort key, this is guaranteed by rules
+			RankRow oldRow = rowKeyMap.get(rowKey);
+			BaseRow oldSortKey = sortKeySelector.getKey(oldRow.row);
+			if (!oldSortKey.equals(sortKey)) {
+				// remove old sort key
+				buffer.remove(oldSortKey, rowKey);
+				// add new sort key
+				int size = buffer.put(sortKey, rowKey);
+				rowKeyMap.put(rowKey, new RankRow(inputRowSer.copy(inputRow), size, true));
+				// update inner rank of records under the old sort key
+				updateInnerRank(oldSortKey);
+			} else {
+				// row content may change, so we need to update row in map
+				rowKeyMap.put(rowKey, new RankRow(inputRowSer.copy(inputRow), oldRow.innerRank, true));
+			}
+			// row content may change, so a retract is needed
+			retract(out, oldRow.row, oldRow.innerRank);
+			collect(out, inputRow);
+		} else if (checkSortKeyInBufferRange(sortKey, buffer)) {
+			// it is an unique record but is in the topN, insert sort key into buffer
+			int size = buffer.put(sortKey, rowKey);
+			rowKeyMap.put(rowKey, new RankRow(inputRowSer.copy(inputRow), size, true));
+			collect(out, inputRow);
+			// remove retired element
+			if (buffer.getCurrentTopNum() > rankEnd) {
+				BaseRow lastRowKey = buffer.removeLast();
+				if (lastRowKey != null) {
+					RankRow lastRow = rowKeyMap.remove(lastRowKey);
+					dataState.remove(lastRowKey);
+					// always send a retraction message
+					delete(out, lastRow.row);
+				}
+			}
+		}
+	}
+
+	private void synchronizeState(Map<BaseRow, RankRow> curRowKeyMap) throws Exception {
```

Comment:
The method name is a little misleading. I think `flushBufferToState(buffer)` or something else is better.